static/js/lib/auth.js

```diff
@@ -44,3 +47,27 @@ export const fetchWithAuthFailure = async (...args: any) => {
     return fetchJSONWithCSRF(...args)
   }
 }
+
+export const fetchWithAuthFailure = async (...args: any) => {
```

Comment:
Suggestion: refactor this and the above to something like this:

```js
const withAuthFailure = (fetchFunc: Function) = {
  return async (...args: any) => {
    try {
      return await fetchFunc(...args)
     } catch (fetchError) {
       if (!fetchError || fetchError.errorStatusCode !== 401) {
         // not an authentication failure, rethrow
         throw fetchError
       }
     
       try {
         // renew the session
         const session = await renewSession()
         const json = await session.json()
         if (!json.has_token) {
           return redirectAndReject("New token was not created")
         }
       } catch (_) {
         return redirectAndReject("Could not renew session")
       }
         // try again now that we have a proper auth
         return fetchFunc(...args)
       }
     }
  }
}

const fetchJSONWithAuthFailure = withAuthFailure(fetchJSONWithCSRF)
const fetchWithAuthFailure = withAuthFailure(fetchWithCSRF)
```

Reply:
That was my first instinct as well, I ended up just going with two separate functions because I ran into an issue trying to mock out things for tests with the above approach. Basically, because if we declare

```js
const fetchJSONWithAuthFailure = withAuthFailure(fetchJSONWithCSRF)
```

Then `fetchJSONWithAuthFailure` has a reference to `fetchJSONWithCSRF` which resolves at module-load time. When we're running the tests we use sinon to stub out `fetchJSONWithCSRF`, but because `fetchJSONWithAuthFailure` is already holding on to the un-stubbed function the stub doesn't do anything.

One way to do something to DRY up the code would be to try to resolve the `fetchFunc` at call-time instead of module-load time. I considered doing something like this:

```js
const withAuthFailure = (fetchJSON: boolean) = {
  return async (...args: any) => {
    const fetchFunc = fetchJSON
      ? fetchJSONWithCSRF
      : fetchWithCSRF
      
    try {
      return await fetchFunc(...args)
     } catch (fetchError) {
       if (!fetchError || fetchError.errorStatusCode !== 401) {
         // not an authentication failure, rethrow
         throw fetchError
       }
     
       try {
         // renew the session
         const session = await renewSession()
         const json = await session.json()
         if (!json.has_token) {
           return redirectAndReject("New token was not created")
         }
       } catch (_) {
         return redirectAndReject("Could not renew session")
       }
         // try again now that we have a proper auth
         return fetchFunc(...args)
       }
     }
  }
}

const fetchJSONWithAuthFailure = withAuthFailure(true)
const fetchWithAuthFailure = withAuthFailure(false)
```

and that works just fine, but idk, it struck me as a big inelegant, so I thought the simplest and most clear thing would be to just go with the two functions. What do you think?

Reply:
Hmm, I agree that's a bit obtuse. An alternative using my original code that should still work with sinon is: 

```js
export const fetchJSONWithAuthFailure = withAuthFailure((...args: any) => fetchJSONWithCSRF(...args))
export const fetchWithAuthFailure = withAuthFailure((...args: any) => fetchWithCSRF(...args))
```

But I'm also ok with things as-is if you feel that extra anonymous func is too much.

Reply:
no I think that makes sense, I'll go with that